### PR TITLE
[Clang] Implement the 'counted_by' attribute

### PR DESCRIPTION
**Pull request by bwendling**
The 'counted_by' attribute is used on flexible array members. The argument for the attribute is the name of the field member in the same structure holding the count of elements in the flexible array. This information can be used to improve the results of the array bound sanitizer and the '__builtin_dynamic_object_size' builtin.

This example specifies the that the flexible array member 'array' has the number of elements allocated for it in 'count':

  struct bar;
  struct foo {
    size_t count;
     /* ... */
    struct bar *array[] __attribute__((counted_by(count)));
  };

This establishes a relationship between 'array' and 'count', specifically that 'p->array' must have *at least* 'p->count' number of elements available. It's the user's responsibility to ensure that this relationship is maintained through changes to the structure.

In the following, the allocated array erroneously has fewer elements than what's specified by 'p->count'. This would result in an out-of-bounds access not not being detected:

  struct foo *p;

  void foo_alloc(size_t count) {
    p = malloc(MAX(sizeof(struct foo),
                   offsetof(struct foo, array[0]) + count *
                       sizeof(struct bar *)));
    p->count = count + 42;
  }

The next example updates 'p->count', breaking the relationship requirement that 'p->array' must have at least 'p->count' number of elements available:

  struct foo *p;

  void foo_alloc(size_t count) {
    p = malloc(MAX(sizeof(struct foo),
                   offsetof(struct foo, array[0]) + count *
                       sizeof(struct bar *)));
    p->count = count + 42;
  }

  void use_foo(int index) {
    p->count += 42;
    p->array[index] = 0; /* The sanitizer cannot properly check this access */
  }

Reviewed By: nickdesaulniers, aaron.ballman

Differential Revision: https://reviews.llvm.org/D148381

### CODE REVIEW

**Comment by bwendling** [comment]
This is identical to the original change (9a954c693573281407f6ee3f4eb1b16cc545033d), but with a fix for the issue @alexfh discovered (https://github.com/llvm/llvm-project/commit/9a954c693573281407f6ee3f4eb1b16cc545033d#commitcomment-129529574). The change is on lines 473-475 of clang/lib/AST/DeclBase.cpp: using `dyn_cast_if_present` instead of just `dyn_cast`. Also added the testcase to `clang/test/CodeGen/bounds-checking.c`.

**Comment by bwendling** [comment]
And @AaronBallman, I haven't forgotten my promise to change the diagnostics. I've been on vacation, and it's become trickier to get them to work the other way than I had first thought. The change is coming soon though.

**Review by oskarwirga — COMMENTED** [review]
I've reviewed the admittedly limited sections I'm familiar with and LGTM! This is great work :)

**Comment by bwendling** [comment]
@AaronBallman I just added a patch that modifies the diagnostics to be more in line with what you wanted. PTAL.

File: clang/include/clang/Sema/TypoCorrection.h [hunk 1]
```diff
@@ -282,7 +282,7 @@ class CorrectionCandidateCallback {
 public:
   static const unsigned InvalidDistance = TypoCorrection::InvalidDistance;
 
-  explicit CorrectionCandidateCallback(IdentifierInfo *Typo = nullptr,
+  explicit CorrectionCandidateCallback(const IdentifierInfo *Typo = nullptr,
```
Comment (AaronBallman) [inline clang/include/clang/Sema/TypoCorrection.h:285]: Improved const correctness -- I love it!! :-) (If the review takes a while, feel free to land the const correctness changes as an NFC commit.)

**Review by AaronBallman — APPROVED** [review]
LGTM! Thank you for the cleanup on the diagnostics, too. :-)

**Comment by bwendling** [comment]
@nickdesaulniers and @alexfh Any comments? :-)

**Review by nickdesaulniers — APPROVED** [review]

**Pull request merged**